libsoftwarecontainer/src/gateway/dbus/dbusgateway.h

```diff
@@ -21,98 +20,50 @@
 #pragma once
 
 #include "gateway.h"
+#include "dbusgatewayinstance.h"
 
-/**
- * @brief DBus Gateway takes care of spawning and killing the DBus proxies.
- *
- *  This module requires the 'dbus-proxy' binary to be available in the $PATH
- *  of the user executing softwarecontainer.
- */
-class DBusGateway :
-    public Gateway
+class DBusGateway : public Gateway
 {
     LOG_DECLARE_CLASS_CONTEXT("DBUS", "DBus gateway");
 
 public:
-    enum ProxyType { SessionProxy, SystemProxy };
 
     static constexpr const char *ID = "dbus";
+    DBusGateway(const std::string &gatewayDir, const std::string &name);
+    virtual ~DBusGateway();
 
     /**
-     * @brief Spawn the proxy and use the supplied path for the socket
-     *
-     * @param type SessionProxy or SystemProxy
+     * @brief Sets config for both dbus session instances
      */
-    DBusGateway(ProxyType type, const std::string &gatewayDir, const std::string &name);
-    virtual ~DBusGateway();
-
-    virtual ReturnCode readConfigElement(const json_t *element) override;
+    virtual ReturnCode setConfig(const std::string &config) override;
 
     /**
-     *  @brief Implements Gateway::activateGateway
-     *
-     *  Starts the dbus-proxy binary and feeds it the configuration set in
-     *  setConfig(). This function will also set up the correct environment
-     *  variables (DBUS_SESSION_BUS_ADDRESS when proxying the session bus, or
-     *  DBUS_SYSTEM_BUS_ADDRESS when proxying the system bus).
-     *
-     *  This function requires a config to have been set using setConfig().
-     *
-     *  This function requires the 'dbus-proxy' binary to be available in $PATH
-     *
-     *  @return true when dbus-proxy was correctly execute and environment
-     *   variable was correctly set
-     *   false if dbus-proxy failed to execute or accept input on STDIN,
-     *   or when environment variable could not be set.
+     * @brief Activates both dbus session instances
      */
-    virtual bool activateGateway();
+    virtual ReturnCode activate() override;
 
     /**
-     * @brief Implements Gateway::teardownGateway
-     *
-     *  This function will clean up processes launched and file descriptors
-     *  opened during the lifetime of the gatway. Specifically it will close
-     *  the connection to the dbus-proxy, and close the stdin and stdout pipes
-     *  opened to dbus-proxy.
-     *
-     * @return false if dbus-proxy could not be terminated, if stdin or stdout
-     *          could not be closed, or if the socket created by the dbus-proxy
-     *          could not be removed.
+     * @brief Stubbed since this class only does containment
      */
-    virtual bool teardownGateway() override;
-
-private:
-    std::string socketName();
-    bool isSocketCreated() const;
-
-    // Socket used for exposing D-Bus in container
-    std::string m_socket;
+    virtual ReturnCode readConfigElement(const json_t *) override { return ReturnCode::SUCCESS; }
```

Comment:
Since you redefine readConfigElement in the cpp file anyway, why have the default implementation like this?